homeassistant/components/downloader/__init__.py

```diff
@@ -57,122 +86,143 @@ def setup(hass: HomeAssistant, config: ConfigType) -> bool:
+                vol.Optional(ATTR_OVERWRITE, default=False): cv.boolean,
+            }
+        ),
+    )
 
-                else:
-                    if filename is None and "content-disposition" in req.headers:
-                        match = re.findall(
-                            r"filename=(\S+)", req.headers["content-disposition"]
-                        )
+    return True
 
-                        if match:
-                            filename = match[0].strip("'\" ")
 
-                    if not filename:
-                        filename = os.path.basename(url).strip()
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload downloader."""
+    hass.services.async_remove(domain=DOMAIN, service=SERVICE_DOWNLOAD_FILE)
+    return True
 
-                    if not filename:
-                        filename = "ha_download"
 
-                    # Check the filename
-                    raise_if_invalid_filename(filename)
+async def _async_download_file(
+    hass: HomeAssistant, service: ServiceCall, download_path: str
+) -> None:
+    """Start thread to download file specified in the URL."""
+    threading.Thread(
+        target=_async_do_download, args=(hass, service, download_path)
+    ).start()
+
+
+def _async_do_download(hass: HomeAssistant, service: ServiceCall, download_path: str):
+    """Download the file."""
+    try:
+        url = service.data.get(ATTR_URL)
+        subdir = service.data.get(ATTR_SUBDIR)
+        filename = service.data.get(ATTR_FILENAME)
+        overwrite = service.data.get(ATTR_OVERWRITE)
+
+        if subdir:
+            # Check the path
+            raise_if_invalid_path(subdir)
+
+        final_path = None
+
+        req = requests.get(url, stream=True, timeout=10)  # type: ignore[arg-type]
+
+        if req.status_code != HTTPStatus.OK:
+            _LOGGER.warning(
+                "Downloading '%s' failed, status_code=%d", url, req.status_code
+            )
+            hass.bus.fire(
+                f"{DOMAIN}_{DOWNLOAD_FAILED_EVENT}",
+                {"url": url, "filename": filename},
+            )
+
+        else:
+            if filename is None and "content-disposition" in req.headers:
+                match = re.findall(
+                    r"filename=(\S+)", req.headers["content-disposition"]
+                )
+                if match:
+                    filename = match[0].strip("'\" ")
 
-                    # Do we want to download to subdir, create if needed
-                    if subdir:
-                        subdir_path = os.path.join(download_path, subdir)
+            if not filename:
+                filename = os.path.basename(url).strip()  # type: ignore[arg-type]
 
-                        # Ensure subdir exist
-                        os.makedirs(subdir_path, exist_ok=True)
+            if not filename:
+                # Generate a random filename
+                filename = "ha_download_" + str(time.time())
```

Comment:
Why is this changed? Please do it in a separate PR instead

Reply:
Basically to identify it originates from HA. It can indeed be done in a separate PR, whereas this PR should focus on the config flow only.